lib/commons/color/get-background-color.js

```diff
@@ -1,132 +1,240 @@
+
+	// rect outside of viewport
+	axe.commons.color.incompleteData.set('bgColor', 'outsideViewport');
+	return null;
 }
+
 /**
- * Calculate alpha transparency of a background element obscuring the current node
- * @private
- * @param {Number} elmIndex
- * @param {Array} elmStack
- * @param {Element} originalElm
- * @return {Number|undefined}
+ * Get relevant stacks of block and inline elements, excluding line breaks
+ * @method getRectStack
+ * @memberof axe.commons.color
+ * @param {Element} elm
+ * @return {Array}
  */
-function calculateObscuringAlpha(elmIndex, elmStack, originalElm) {
-	var totalAlpha = 0;
+color.getRectStack = function(elm) {
+	const boundingCoords = axe.commons.color.centerPointOfRect(
+		elm.getBoundingClientRect()
+	);
 
-	if (elmIndex > 0) {
-		// there are elements above our element, check if they contribute to the background
-		for (var i = elmIndex - 1; i >= 0; i--) {
-			let bgElm = elmStack[i];
-			let bgElmStyle = window.getComputedStyle(bgElm);
-			let bgColor = getBgColor(bgElm, bgElmStyle);
-			if (bgColor.alpha && contentOverlapping(originalElm, bgElm)) {
-				totalAlpha += bgColor.alpha;
-			} else {
-				// remove elements not contributing to the background
-				elmStack.splice(i, 1);
+	if (!boundingCoords) {
+		return null;
+	}
+
+	let boundingStack = dom.shadowElementsFromPoint(
+		boundingCoords.x,
+		boundingCoords.y
+	);
+
+	let rects = Array.from(elm.getClientRects());
+	// If the element does not have multiple rects, like for display:block, return a single stack
+	if (!rects || rects.length <= 1) {
+		return [boundingStack];
+	}
+
+	// Handle inline elements spanning multiple lines to be evaluated
+	let filteredArr = rects
+		.filter(rect => {
+			// exclude manual line breaks in Chrome/Safari
+			return rect.width && rect.width > 0;
+		})
+		.map(rect => {
+			const coords = axe.commons.color.centerPointOfRect(rect);
+			if (coords) {
+				return dom.shadowElementsFromPoint(coords.x, coords.y);
 			}
-		}
+		});
+
+	if (filteredArr.some(stack => stack === undefined)) {
+		// Can be happen when one or more of the rects sits outside the viewport
+		return null;
 	}
-	return totalAlpha;
-}
+
+	// add bounding client rect stack for comparison later
+	filteredArr.splice(0, 0, boundingStack);
+	return filteredArr;
+};
+
 /**
- * Determine if element is partially overlapped, triggering a Can't Tell result
+ * Look at document and body elements for relevant background information
+ * @method sortPageBackground
  * @private
- * @param {Element} elm
- * @param {Element} bgElm
- * @param {Object} bgColor
- * @return {Boolean}
+ * @param {Array} elmStack
+ * @returns {Array}
  */
-function elmPartiallyObscured(elm, bgElm, bgColor) {
-	var obscured =
-		elm !== bgElm && !dom.visuallyContains(elm, bgElm) && bgColor.alpha !== 0;
-	if (obscured) {
-		axe.commons.color.incompleteData.set('bgColor', 'elmPartiallyObscured');
+function sortPageBackground(elmStack) {
+	let bodyIndex = elmStack.indexOf(document.body);
```

Comment:
I started to look at this method on #1456 as it did not cover a specific use case. `body` was not in the background stack since there's cases where an element and `body` do not intersect, but still has a background color due to `body` filling the visible canvas. Since this is tackling the refactoring of background color, it might be good to handle that case here.

Reply:
@scurker I think the fix to tackle the linked issue should be done in a different PR. This PR was purely moving existing code around to make things easier for some future work.